src/v/storage/segment_appender.cc

```diff
@@ -460,6 +502,38 @@ void segment_appender::dispatch_background_head_write() {
 
 ss::future<> segment_appender::flush() {
     _inactive_timer.cancel();
+
+    // dispatched write will drive flush completion
+    if (_head && _head->bytes_pending()) {
+        auto& w = _flush_ops.emplace_back(file_byte_offset());
+        dispatch_background_head_write();
+        return w.p.get_future();
+    }
+
+    if (file_byte_offset() <= _flushed_offset) {
+        return ss::now();
+    }
+
+    /*
+     * if there are inflight write _ops_ then we can be sure that flush ops will
+     * be processed eventually (see: maybe advance stable offset).
+     */
+    if (!_inflight.empty()) {
+        auto& w = _flush_ops.emplace_back(file_byte_offset());
+        return w.p.get_future();
+    }
+
+    vassert(file_byte_offset() <= _stable_offset,
+        "No inflight writes but eof {} > stable offset {}",
+        file_byte_offset(), _stable_offset);
```

Comment:
in think all the vasserts() here should print `*this`  to track metadata errors.